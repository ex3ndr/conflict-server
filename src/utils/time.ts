import { events } from "../app/events";

export async function delay(ms: number) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

export function exponentialBackoffDelay(currentFailureCount: number, minDelay: number, maxDelay: number, maxFailureCount: number) {
    let maxDelayRet = minDelay + ((maxDelay - minDelay) / maxFailureCount) * Math.max(currentFailureCount, maxFailureCount);
    return Math.round(Math.random() * maxDelayRet);
}

export type BackoffFunc = <T>(callback: () => Promise<T>) => Promise<T>;

export function createBackoff(
    opts?: {
        onError?: (e: any, failuresCount: number) => void,
        minDelay?: number,
        maxDelay?: number,
        maxFailureCount?: number
    }): BackoffFunc {
    return async <T>(callback: () => Promise<T>): Promise<T> => {
        let currentFailureCount = 0;
        const minDelay = opts && opts.minDelay !== undefined ? opts.minDelay : 250;
        const maxDelay = opts && opts.maxDelay !== undefined ? opts.maxDelay : 1000;
        const maxFailureCount = opts && opts.maxFailureCount !== undefined ? opts.maxFailureCount : 50;
        while (true) {
            try {
                return await callback();
            } catch (e) {
                if (currentFailureCount < maxFailureCount) {
                    currentFailureCount++;
                }
                if (opts && opts.onError) {
                    opts.onError(e, currentFailureCount);
                }
                let waitForRequest = exponentialBackoffDelay(currentFailureCount, minDelay, maxDelay, maxFailureCount);
                await delay(waitForRequest);
            }
        }
    };
}

export let backoff = createBackoff({ onError: (e) => { console.warn(e); } });

export function delayBreakable(ms: number) {
    // We can cancel delay from outer code
    let promiseResolver: ((value?: any | PromiseLike<any>) => void) | null = null;
    let resolver = () => {
        if (promiseResolver) {
            promiseResolver();
        }
    };
    let promise = new Promise(resolve => {
        promiseResolver = resolve;
        setTimeout(resolve, ms);
    });
    return { promise, resolver };
}

export function resumeKey(key: string) {
    events.emit('pause-' + key);
}

export async function pauseWithKey(ms: number, key: string) {
    let { promise, resolver } = delayBreakable(ms);
    events.addListener('pause-' + key, resolver);
    try {
        await promise;
    } finally {
        events.removeListener('pause-' + key, resolver);
    }
}