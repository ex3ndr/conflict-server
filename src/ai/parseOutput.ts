import { trimIndent } from "../utils/text";

export function parseOutput(src: string) {

    src = trimIndent(src);

    // Remove tail and head
    if (src.startsWith('#PUBLIC')) {
        src = trimIndent(src.substring('#PUBLIC'.length));
    }
    if (src.endsWith('#END')) {
        src = trimIndent(src.substring(0, src.length - '\n#END'.length));
    }

    let publicMessage: string | null = null;
    let secretA: string | null = null;
    let secretB: string | null = null;

    // Extract public message
    let n = src.indexOf('#SECRET_A');
    if (n > -1) {
        let tt = trimIndent(src.substring(0, n));
        if (tt.length > 0) {
            publicMessage = tt;
        }
        src = trimIndent(src.substring(n + '#SECRET_A'.length));
    } else {
        if (src.length > 0) {
            publicMessage = src;
        }
        src = '';
    }

    // Extract secret message
    n = src.indexOf('#SECRET_B');
    if (n > -1) {
        let tt = trimIndent(src.substring(0, n));
        if (tt.length > 0) {
            secretA = tt;
        }
        src = trimIndent(src.substring(n + '#SECRET_B'.length));
    } else {
        if (src.length > 0) {
            secretA = src;
        }
        src = '';
    }
    if (src.length > 0) {
        secretB = src;
    }

    return {
        publicMessage,
        secretA,
        secretB
    };
}