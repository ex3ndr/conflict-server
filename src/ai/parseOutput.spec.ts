import { trimIndent } from "../utils/text";
import { parseOutput } from "./parseOutput";

describe('parseOutput', () => {
    it('should parse output', () => {
        let parsed = parseOutput(trimIndent(`
        #PUBLIC
        pppp
        #SECRET_A
        #SECRET_B
        #END
        `));
        expect(parsed).toMatchSnapshot();

        parsed = parseOutput(trimIndent(`
        #PUBLIC
        #SECRET_A
        A
        #SECRET_B
        #END
        `));
        expect(parsed).toMatchSnapshot();

        parsed = parseOutput(trimIndent(`
        #PUBLIC
        #SECRET_A
        A
        #SECRET_B
        B
        #END
        `));
        expect(parsed).toMatchSnapshot();

        parsed = parseOutput(trimIndent(`
        #PUBLIC
        #SECRET_A
        #SECRET_B
        B
        #END
        `));
        expect(parsed).toMatchSnapshot();
    });
});