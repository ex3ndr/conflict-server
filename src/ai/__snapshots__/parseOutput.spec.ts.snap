// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseOutput should parse output 1`] = `
{
  "publicMessage": "pppp",
  "secretA": null,
  "secretB": null,
}
`;

exports[`parseOutput should parse output 2`] = `
{
  "publicMessage": null,
  "secretA": "A",
  "secretB": null,
}
`;

exports[`parseOutput should parse output 3`] = `
{
  "publicMessage": null,
  "secretA": "A",
  "secretB": "B",
}
`;

exports[`parseOutput should parse output 4`] = `
{
  "publicMessage": null,
  "secretA": null,
  "secretB": "B",
}
`;
